File: .github/checks/wasm-ci-paths/check.mjs
import { execFileSync } from "node:child_process";
import { existsSync, readFileSync, readdirSync } from "node:fs";
import { dirname, join, relative } from "node:path";

const abort = (message) => {
    console.error(message);
    process.exit(1);
};
const readJSON = (path) => JSON.parse(readFileSync(path, "utf8"));

process.chdir(process.argv[2] ?? ".");
const root = process.cwd();
const pathsFile = ".github/scripts/ci/paths.json";
const patterns = readJSON(pathsFile)["web-lint"];
if (patterns === null) process.exit(0);
if (patterns.some((pattern) => pattern.startsWith("!"))) abort(`${pathsFile}: WASM coverage does not support negated paths`);

const { packages } = JSON.parse(execFileSync("cargo", [
    "metadata", "--no-deps", "--format-version", "1", "--locked", "--offline",
], { cwd: "rust", encoding: "utf8" }));
const byName = new Map(packages.map((pkg) => [pkg.name, pkg]));
const byDirectory = new Map(packages.map((pkg) => [dirname(pkg.manifest_path), pkg]));
const pending = [];
for (const entry of readdirSync("web/packages/wasm")) {
    const path = join("web/packages/wasm", entry, "package.json");
    if (entry.startsWith(".") || !existsSync(path)) continue;
    const { name } = readJSON(path);
    const pkg = byName.get(name);
    if (!pkg) abort(`${path}: no matching Cargo package for ${name}`);
    pending.push(dirname(pkg.manifest_path));
}
if (!pending.length) abort("No Web WASM packages found");

const directories = new Set();
while (pending.length) {
    const directory = pending.pop();
    if (directories.has(directory)) continue;
    directories.add(directory);
    const pkg = byDirectory.get(directory);
    if (!pkg) abort(`Local dependency ${directory} is outside the Rust workspace`);
    // Include optional and target-specific edges without fetching Cargo's resolved graph.
    for (const dependency of pkg.dependencies) {
        if (dependency.kind !== "dev" && dependency.path) pending.push(dependency.path);
    }
}

const required = ["rust/Cargo.lock", "rust/Cargo.toml", ...[...directories].map((dir) => `${relative(root, dir)}/**`)];
// Literal directory prefixes prove coverage of future files too.
const prefixes = patterns.filter((pattern) => /^[\w./-]+\/\*\*$/.test(pattern)).map((pattern) => pattern.slice(0, -2));
const missing = required.filter((path) => !patterns.includes(path) && !prefixes.some((prefix) => path.startsWith(prefix)));
for (const path of missing.sort()) console.error(`${pathsFile}: missing WASM dependency path ${JSON.stringify(path)}`);
process.exit(missing.length ? 1 : 0);

File: .github/checks/wasm-ci-paths/check.rb
require "json"
require "open3"
require "pathname"
require "set"
require "yaml"

Dir.chdir(ARGV.fetch(0, "."))
root = Pathname.pwd
workflow_path = ".github/workflows/web-lint.yml"
workflow = YAML.safe_load(File.read(workflow_path), aliases: true)
events = workflow["on"] || workflow[true]
config = events.fetch("pull_request") || {}
abort "#{workflow_path}: WASM coverage does not support paths-ignore" if config.key?("paths-ignore")
patterns = config["paths"]
exit 0 unless patterns
abort "#{workflow_path}: WASM coverage does not support negated paths" if patterns.any? { |p| p.start_with?("!") }

output, status = Open3.capture2(
  "cargo", "metadata", "--no-deps", "--format-version", "1", "--locked", "--offline",
  chdir: "rust",
)
exit status.exitstatus unless status.success?
packages = JSON.parse(output).fetch("packages")
by_name = packages.to_h { |package| [package.fetch("name"), package] }
by_directory = packages.to_h { |package| [Pathname.new(package.fetch("manifest_path")).dirname, package] }
pending = Dir["web/packages/wasm/*/package.json"].map do |path|
  name = JSON.parse(File.read(path)).fetch("name")
  package = by_name.fetch(name) { abort "#{path}: no matching Cargo package for #{name}" }
  Pathname.new(package.fetch("manifest_path")).dirname
end
abort "No Web WASM packages found" if pending.empty?

directories = Set.new
until pending.empty?
  directory = pending.pop
  next unless directories.add?(directory)

  package = by_directory.fetch(directory) { abort "Local dependency #{directory} is outside the Rust workspace" }
  # Include optional and target-specific edges to avoid fetching Cargo's resolved graph.
  package.fetch("dependencies").each do |dependency|
    next if dependency["kind"] == "dev" || !dependency["path"]

    pending << Pathname.new(dependency.fetch("path"))
  end
end

required = %w[rust/Cargo.lock rust/Cargo.toml] + directories.map { |dir| "#{dir.relative_path_from(root)}/**" }
# Literal directory prefixes prove coverage of the entire crate, including future files.
prefixes = patterns.grep(%r{\A[\w./-]+/\*\*\z}).map { |pattern| pattern.delete_suffix("**") }
missing = required.reject { |path| patterns.include?(path) || prefixes.any? { |prefix| path.start_with?(prefix) } }
missing.sort.each { |path| warn "#{workflow_path}: missing WASM dependency path #{path.inspect}" }
exit(missing.empty? ? 0 : 1)
